pyscf_ipu/experimental/nuclear_gradients.py

```diff
@@ -0,0 +1,92 @@
+# Copyright (c) 2023 Graphcore Ltd. All rights reserved.
+from typing import Callable
+import jax.numpy as jnp
+from jax import jit, tree_map, vmap
+from jax.ops import segment_sum
+
+from .basis import Basis
+from .integrals import _overlap_primitives, _kinetic_primitives
+from .orbital import batch_orbitals
+from .primitive import Primitive
+from .types import Float3, Float3xNxN
+
+
+def grad_primitive_integral(
+    primitive_op: Callable, atom_index: int, a: Primitive, b: Primitive
+) -> Float3:
+    """Generic gradient of a one-electron integral with respect the atom_index center"""
+    axes = jnp.arange(3)
+    lhs_p1 = vmap(a.offset_lmn, (0, None))(axes, 1)
+    t1 = 2 * a.alpha * vmap(primitive_op, (0, None))(lhs_p1, b)
+
+    lhs_m1 = vmap(a.offset_lmn, (0, None))(axes, -1)
+    t2 = jnp.where(a.lmn > 0, a.lmn, jnp.zeros_like(a.lmn))
+    t2 *= vmap(primitive_op, (0, None))(lhs_m1, b)
```

Comment:
I realise this is sacrilege, but I wonder if a python loop over 0,1,2 might be easier to read, and ultimately be no slower, as it will be vectorisable by a higher level vmap.
E.g. something like (artist's impression)
```
grad = jnp.zero(3)
for axis in range(3):
  t1 = 2 * a.alpha * primitive_op(a.offset_lmn[axis], 1, b)
  t2 = a.offset_lmn[axis] * primitive_op(a.offset_lmn[axis], -1, b)
  grad = grad.at(axis).set(t1 - t2)

Reply:
Or if that turns out to be slower, it might serve as reference code in another function so testing can compare them.

Reply:
trying out a list comprehension like this:

```python
    t1 = [primitive_op(a.offset_lmn(ax, 1), b) for ax in range(3)]
    t2 = [primitive_op(a.offset_lmn(ax, -1), b) for ax in range(3)]
    grad_out = 2 * a.alpha * jnp.stack(t1) - a.lmn * jnp.stack(t2)
    return grad_out
```
better or worse?

Reply:
Definitely easier to read - perf impact ?